scripts/check-truffle-namespace-dependency-versions.js

```diff
@@ -0,0 +1,87 @@
+const fs = require("fs");
+const path = require("path");
+
+const chalk = require("chalk");
+
+const debug = require("debug")("check-package-versions");
+
+function main() {
+  const packages = readPackages();
+  const errors = [];
+
+  for (const candidateName in packages) {
+    for (const dependencyName in packages) {
+      errors.push(
+        ...checkDependencyVersionRange(
+          packages[candidateName],
+          packages[dependencyName]
+        )
+      );
+    }
+  }
+
+  for (const error of errors) {
+    console.error(chalk.red("Error:"), error);
+  }
+  const doneColor = errors.length > 0 ? chalk.red : chalk.green;
+  console.error(
+    doneColor("Done."),
+    `Found ${errors.length} errors for @truffle namespace version dependencies.`
+  );
+
+  if (errors.length > 0) {
+    process.exit(1);
+  }
+}
+
+function getPackageDirectoryNames() {
+  return fs.readdirSync(path.join(__dirname, "..", "packages"));
+}
+
+function readPackages() {
+  const packages = getPackageDirectoryNames();
+  const packageSpecs = {};
+  for (const packageName of packages) {
+    const packageFilePath = path.join(
+      path.resolve(__dirname, "..", "packages", packageName),
+      "package.json"
+    );
+    const rawJson = fs.readFileSync(packageFilePath, { encoding: "utf8" });
+    packageSpecs[packageName] = JSON.parse(rawJson);
+  }
+
+  return packageSpecs;
+}
+
+const depTypes = [
+  "dependencies",
+  "devDependencies",
+  "peerDependencies",
+  "optionalDependencies"
+];
+
+function* checkDependencyVersionRange(candidate, dependency) {
+  const name = dependency.name;
+  const version = dependency.version;
+
+  for (const depType of depTypes) {
+    const deps = candidate[depType];
+    if (deps && Object.keys(deps).includes(name)) {
```

Comment:
Not a request to change but I'm curious why you chose `Object.keys` over `deps.hasOwnProperty(name)` or the `in` operator here.

Reply:
Ah, that's a better approach. To answer the question - in the moment I preferred `Object.keys` as it will just return an empty list for any type without enumerable properties, including `null` and `undefined`. But the test for a truthy value of `deps` here makes that concern moot, so I can either drop that test, or convert to `deps.hasOwnProperty(...)`.

Reply:
That's your call :)